### PR TITLE
added timesheet display

### PR DESCRIPTION
**Pull request by Krishna-Sivakumar**
Sessions can be visualized as timesheets now.

### In Action
---
image

### Updated Help Section
---
image

Would be better to display the timesheet for the current and previous month only as the database gets bigger, but that's work for the future

### CODE REVIEW

File: src/commands/log/mod.ts [hunk 1]
```diff
@@ -149,6 +154,68 @@ function fmtSession(idx: string, sess: Session, indentWidth: number) {
...
 
+function printTimesheet(sessions: Session[]) {
+    // bucket sessions into days
+
+    let buckets: Map<number, number> = new Map();
+    let table: string[][] = [];
+    sessions.forEach((session) => {
+        let start = new Date(session.start);
+        let end = new Date(session.end ? session.end : Date.now());
+        const [startyear, startmonth, startdate] = destructureDate(start);
```
Comment (xyzshantaram) [inline src/commands/log/mod.ts]: I'd prefer if these variable names were shorter, maybe starty,startm,startd or similar.

File: src/commands/log/mod.ts [hunk 2]
```diff
@@ -149,6 +154,68 @@ function fmtSession(idx: string, sess: Session, indentWidth: number) {
...
+    let current_month = -1;
+
+    buckets.forEach((millis, key) => {
+        let bucket_date = new Date(key)
+        let month_string = "";
+        if (current_month != bucket_date.getMonth() - 1) {
+            current_month = bucket_date.getMonth() - 1;
+            month_string = new Intl.DateTimeFormat("en-GB", {month: "long"}).format(bucket_date)
+        }
+        const minutes  = Math.floor(millis / 1000 / 60);
```
Comment (xyzshantaram) [inline src/commands/log/mod.ts]: Why do this instead of using humanReadable(millis, true)?

File: src/commands/log/mod.ts [hunk 3]
```diff
@@ -149,6 +154,68 @@ function fmtSession(idx: string, sess: Session, indentWidth: number) {
...
+        const updateBucket = (key: number, val: number) => {
+            let prevVal = buckets.get(key) || 0;
+            buckets.set(key, Math.min(prevVal + val, 24 * 60 * 60 * 1000));
+        };
+
+        if (startdate === enddate) {
+            // session does not spill into another day
+            const key = new Date(startyear, startmonth - 1, startdate).getTime();
+            let hours = end.getTime() - start.getTime();
+            updateBucket(key, hours);
```
Comment (xyzshantaram) [inline src/commands/log/mod.ts:177]: Would it be possible to do an early return here to avoid the big else block?

File: src/commands/log/mod.ts [hunk 4]
```diff
@@ -149,6 +154,68 @@ function fmtSession(idx: string, sess: Session, indentWidth: number) {
...
+            }
+        }
+    });
+
+    table.push([heading("Month"), heading("Date"), heading("Hours Logged")]);
+    let current_month = -1;
+
+    buckets.forEach((millis, key) => {
+        let bucket_date = new Date(key)
+        let month_string = "";
```
Comment (xyzshantaram) [inline src/commands/log/mod.ts]: Please avoid snake_case variable names.

**Review by xyzshantaram — CHANGES_REQUESTED** [review]
Really solid PR @Krishna-Sivakumar. I just have a few minor nitpicks here and there for which I've left review comments, but this is basically perfect - I think I'll adopt this table style for the main log as well, not sure why it didn't strike me to do it like this before. Thanks a lot!

File: src/commands/log/mod.ts (hunk 2 above)
Comment (Krishna-Sivakumar) [reply src/commands/log/mod.ts]: makes sense

File: src/commands/log/mod.ts (hunk 4 above)
Comment (Krishna-Sivakumar) [reply src/commands/log/mod.ts]: forgot about this, thanks

File: src/commands/log/mod.ts (hunk 3 above)
Comment (Krishna-Sivakumar) [reply src/commands/log/mod.ts:177]: it keeps the logic explicit and doesn't incur any performance loss so idt it needs to be changed

File: src/commands/log/mod.ts (hunk 1 above)
Comment (Krishna-Sivakumar) [reply src/commands/log/mod.ts]: makes it more readable, so I want to keep this around

Comment (xyzshantaram) [reply src/commands/log/mod.ts]: In that case, please fix the casing, and we should be good to merge!

**Review by xyzshantaram — CHANGES_REQUESTED** [review]

**Comment by xyzshantaram** [comment]
Re: this
> Would be better to display the timesheet for the current and previous month only as the database gets bigger, but that's work for the future

We could split off timesheet into its own subcommand and give it a --since option that accepts a date delta similar to `etu session edit`'s time-editing functionality.

Also, do you think it makes sense to change `etu log --timesheet` to something like `etu log --months`? I'm thinking of making  the default log view output a table as well.

**Comment by Krishna-Sivakumar** [comment]
> We could split off timesheet into its own subcommand and give it a --since option that accepts a date delta similar to etu session edit's time-editing functionality.

I was thinking of adding a subcommand with more controls as well, if this goes in the direction of providing more control over what part of the timesheet to display.

> Also, do you think it makes sense to change etu log --timesheet to something like etu log --months? I'm thinking of making the default log view output a table as well.

If it's gonna be separated later on, I think we can remove the option altogether. But I do think filtering the existing logs with a year,month,week format might be nice! How exactly to do this, Isn't something I have an idea about.

**Review by xyzshantaram — APPROVED** [review]
LGTM!

**Pull request merged**